[FLINK-15802][table] Support new type inference for table functions

twalthr opened the pull request:
## What is the purpose of the change

Enables the new type inference for table functions. The PR conists of two commits.

The first commit refactors the code generation around table functions such that return data type and argument data type are read from the function at the same location instead of reading each in a separate code generator. This adds an additional runtime collector that is only responsible for result conversion. All existing tests pass after this change.

The second commit further develops the stack based on `BridgingSqlFunction` and the new type inference. Similar to FLINK-15487, we don't test every data type combination yet.

## Brief change log

See commit messages.

## Verifying this change

This change is already covered by existing tests, such as `CorrelateITCase`.

Additional tests added in `FunctionITCase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

twalthr commented:
CC @JingsongLi

twalthr commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/sql/validate/ProcedureNamespace.java:
```diff
@@ -0,0 +1,80 @@
+ * Namespace whose contents are defined by the result of a call to a user-defined procedure.
+ *
+ * <p>Note: Compared to Calcite, this class implements custom logic for dealing with collection tables
+ * like {@code TABLE(function(...))} procedures. Compared to the SQL standard, Flink's table functions
```
The `LATERAL` keyword does not matter here. `TABLE` is defined in the SQL standard as a function that converts a functions output into a table. The function can take literals, expressions, etc.

JingsongLi replied:
But no input reference fields?

twalthr replied:
Input fields might depend on a `LATERAL` keyword. But in any case, this JavaDoc and class is correctly describing the SQL standard "collection table".

JingsongLi replied:
OK, I got the meaning of "collection table".

twalthr commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/TableFunctionCallGen.scala:
```diff
@@ -49,7 +71,7 @@ class TableFunctionCallGen(tableFunction: TableFunction[_]) extends CallGenerato
     // has no result
     GeneratedExpression(
-      functionReference,
+      resultCollectorTerm,
```
Yes, the logic was also complicated before. Conceptually, I think returning a collector as a result is better than the function itself. But I agree, I will add more documentation to the class.

twalthr commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/TableFunctionCallGen.scala:
```diff
@@ -60,7 +82,61 @@ class TableFunctionCallGen(tableFunction: TableFunction[_]) extends CallGenerato
+    val externalTerm = newName("externalRecord")
+    val resultGenerator = new ExprCodeGenerator(collectorCtx, true)
+      .bindInput(externalType, externalTerm, pojoFieldMapping)
+    val wrappedResult = resultGenerator.generateConverterResultExpression(
```
good point, I changed the logic again. So we will skip the conversion for structured types. This also means that we have different handling for `collect(null)`. I think we should skip emitting in those cases.

twalthr commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/collector/WrappingCollector.java:
```diff
@@ -0,0 +1,49 @@
+	 * Outputs the final result to the wrapped collector.
+	 */
+	public void outputResult(T result) {
+		this.collector.collect(result);
```
This is what I also explained in the PR description. We will have another collector indirection. But the JVM should be smart enough to remove this simply call. I smaller projection collector is also beneficial for inlining the method. So I don't expect performance regressions here.

twalthr commented:
Thanks for the feedback @JingsongLi and @dawidwys. I addressed your comments. I could further improve the runtime code in the collector. I also documented the behavior for `collect(null)`.

twalthr commented:
@flinkbot run travis

twalthr commented:
@flinkbot run travis

JingsongLi commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/sql/validate/ProcedureNamespace.java:
```diff
@@ -0,0 +1,80 @@
+		final SqlCallBinding callBinding =
+			new SqlCallBinding(validator, scope, call);
+		// legacy table functions
+		if (operator instanceof TableSqlFunction) {
```
It seems that we can use `TableSqlFunction`, because `TableSqlFunction`s are different between legacy planner and blink planner.
This lead to we have different codes in between legacy planner and blink planner.

twalthr replied:
I will use the previous `instanceof SqlUserDefinedTableFunction` to avoid any classloading issues. I copied the `ProcedureNamespace` in both planners so we should not have problems.

twalthr commented:
@flinkbot run travis

twalthr commented:
Travis gave green light. Merging...
